crates/shared/src/price_estimation/trade_verifier.rs

```diff
@@ -400,16 +463,89 @@ fn encode_settlement(
         &query.in_amount.get(),
     );
 
-    EncodedSettlement {
+    let mut trades = vec![encoded_trade];
+
+    if let Trade::WithJitOrders(trade) = trade {
+        for jit_order in trade.jit_orders.iter() {
+            let order_data = OrderData {
+                sell_token: jit_order.sell_token,
+                buy_token: jit_order.buy_token,
+                receiver: Some(jit_order.receiver),
+                sell_amount: jit_order.sell_amount,
+                buy_amount: jit_order.buy_amount,
+                valid_to: jit_order.valid_to,
+                app_data: jit_order.app_data,
+                fee_amount: 0.into(),
+                kind: match &jit_order.side {
+                    dto::Side::Buy => OrderKind::Buy,
+                    dto::Side::Sell => OrderKind::Sell,
+                },
+                partially_fillable: false,
+                sell_token_balance: jit_order.sell_token_source,
+                buy_token_balance: jit_order.buy_token_destination,
+            };
+            let (owner, signature) = match jit_order.signing_scheme {
+                SigningScheme::Eip1271 => {
+                    let (owner, signature) = jit_order.signature.split_at(20);
+                    let owner = H160::from_slice(owner);
+                    let signature = Signature::from_bytes(jit_order.signing_scheme, signature)?;
+                    (owner, signature)
+                }
+                SigningScheme::PreSign => {
+                    let owner = H160::from_slice(&jit_order.signature);
+                    let signature =
+                        Signature::from_bytes(jit_order.signing_scheme, Vec::new().as_slice())?;
+                    (owner, signature)
+                }
+                _ => {
+                    let signature =
+                        Signature::from_bytes(jit_order.signing_scheme, &jit_order.signature)?;
+                    let owner = signature
+                        .recover(domain_separator, &order_data.hash_struct())?
+                        .context("could not recover the owner")?
+                        .signer;
+                    (owner, signature)
+                }
+            };
+
+            tokens.push(jit_order.sell_token);
```

Comment:
Now that we can settle everything at the UCP I think we should prepare the price vector once with all the prices reported by the solver and only look up the index here rather than pushing duplicate prices.

Reply:
Given that the price vector will always be very small (<10) I think it would be totally fine to do a simple linear search through the vector to find the index.